File: 814-binary-tree-pruning/814-binary-tree-pruning.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pruneTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        def delete(root):
            if not root:
                return True
            if root.val == 0 and not root.left and not root.right:
                return True
            left, right = delete(root.left), delete(root.right)
            if left:
                root.left = None
            if right:
                root.right = None
            if left and right and root.val == 0:
                return True
            return False
        sanitel = TreeNode()
        sanitel.left = root
        delete(sanitel)
        return sanitel.left
                
        

File: 814-binary-tree-pruning/814-binary-tree-pruning.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode pruneTree(TreeNode root) {
        TreeNode dumb =new TreeNode(1);
        dumb.left = root;
        TreeNode res = prune(dumb);
        return dumb.left;
        
    }
    //true 
    public TreeNode prune(TreeNode root){
        if(root == null)return null;
        TreeNode left = prune(root.left);
        TreeNode right = prune(root.right);
        if(root.val == 0 && left == null && right == null)return null;
        root.left = left;
        root.right = right;
        return root;
    }
}